### ISSUE TITLE
Question/Feature Request: Default Reviewers Based on Repo

### ISSUE THREAD

**Issue by marco-tranzatto** [refined-bitbucket/refined-bitbucket#201]
Hi all,

first of all thanks a lot for this amazing extension! I came across the link to this repo while looking how to add default pull request template in bitbucket ... I already like a lot my new browser extension!

I'm currently working in a team of more than 20 developers, mainly using bitbucket, and something we are missing a lot is the possibility to add default reviews on each repo. Let me be more specific: we would like to have a list of groups of reviewers, and assign each repo we have to one (or multiple) group of reviewers.

For example, there could be one file specifying groups:
- reviewers_communication_repos: person1, person3, person5
- reviwers_control_repos: person1, person10
etc.

Then each repo is associated to one or more groups. When a new pull requests is opened then the reviewers of that pull requests are automatically taken from the specified group(s).

What do you think about this? Is it double or rather difficult to implement? Could a person without any experience in javascript (like me) implement this feature?
I look forward to hearing your opinion.
Regards,
Marco.

**Comment by reyronald** [comment]
Hello @marco-tranzatto !

Thank you for your kind words! User satisfaction and that kind of feedback is what motivates me to keep maintaining this extension!

Regarding your feature request, I'm not sure I understand how it would be different than what Bitbucket already offers. There's an option under a repo's settings called Default Reviewers:

![default-reviewers](https://user-images.githubusercontent.com/7514993/38810862-06449efe-4156-11e8-8eae-1ce00510ffcb.png)

My team and myself are already using this feature in all of our repos and from what I could understand it works like you described. Could you elaborate on how the feature you propose would be different?

**Comment by marco-tranzatto** [comment]
Hi @reyronald , thanks for your fast reply!

Probably I didn't explain myself completely. Let me try again :) 
I'm aware of the built-in function of Bitbucket, but from our point of view it's just not enough. We have more than 150 repositories and setting default reviewers for each of them is not only time consuming but even quite useless on the long term run. A more easy and scalable solution, in our opinion, would be to specify for each repo a group of maintainers. These groups are specified in a single "configuration" file stored on Bitbucket. If Person1 belongs to GroupA and  either leaves the company  or simply is assigned to take care of some other repos, then we would need to change only the global "config" file: we delete Person1 from GroupA and add him/her to GroupC and GroupD, for example. We would then not need at all to reflect this change repo by repo as it would be handle directly by your plugin. This feature would be extremely helpful even because many of our repos are "logically" linked and they have the same group of maintainers. 

I hope I could make it a little bit more clear now. What to dou think about it?
Regards,
Marco.

**Comment by reyronald** [comment]
Oh I see! I get it now.

You mentioned that the configuration file would be stored in Bitbucket, so you mean you would like to include this file in the repo? If that's the case then you would have to go commit that file in your 150 repos and I'm assuming the configuration would only be applicable to the current repo, so then the contents of those 150 files themselves would have to be different per case. Is that what you had in mind?

Also, could you post an example of how you visualized the contents of this file to be?

**Comment by marco-tranzatto** [comment]
Hi @reyronald 

You're somehow correct, but in my vision there exists two files: one unique file stored in one "configuration_repo" in Bitbucket that lists all the groups and people belonging to these groups, and one file for each repo, that indicates to which groups the repo is linked to.

For the first file (global file indicating groups and people) I had in mind something similar to what you already offer for the pull request template: put a URL to this single (and unique) configuration file stored in one repository, hosted in Bitbucket. This file would then be unique for all the repositories and shared among the whole team of developers. 

For the second file each repo should have its own version (with fixed and predefined name) that basically indicates to which reviewer group(s) it's associated with. Since group names don't not change often, it's fine to create a file for each repo, stating to which groups it's connected t. What change are people associated to each group. And so what changes often is only the first file.

An example of the unique (and shared) configuration file would be the following (file `maintainers_list.md` in repo `configuration_repo`) is:

\=======================================
Group1:  \# This is a comment
  \- Bitbucket_Account_Person1 \# This corresponds to Mr. XXX YYY,
  \- Bitbucket_Account_Person2
  \- Bitbucket_Account_Person5 

Group2:  \# This is a comment
  \- Bitbucket_Account_Person1
  \- Bitbucket_Account_Person10
\=======================================

Then in each repo we want to assign groups we would have one single file to express the link:
In repo `my_awesome_repo` there exists the file with default name `groups_maintainers.md`
\=======================================
Group1
Group55
\=======================================


When opening a new pull request in repo `my_awesome_repo`, then the plugin reads the file `my_awesome_repo/groups_maintainers.md`, for each group found in this file in checks the shared file `maintainers_list.md` and fills in the field "Reviewers" with the names associated to that group.
![bit_re](https://user-images.githubusercontent.com/5328930/38925089-e06d4a82-42fe-11e8-954d-f2d1fa1ee674.png)


What do you think about this?

Regards,
Marco.

**Comment by marco-tranzatto** [comment]
@reyronald, did you by chance have time to have a look at the previous post? I'm curious to know what you think about it :)

Cheers,
Marco.

**Comment by requaos** [comment]
I look forward to maintaining my set of default reviewers, per repo, independent of the default reviewers specified in the repository settings. The first reason being that the repository maintainer is slow to update this list when turnover occurs and the second being that I do not have permissions to access the settings of the repo. Excellent plugin, and a shoutout to @drexler for Vagrantfile and Jenkinsfile support of syntax highlighting

**Comment by reyronald** [comment]
Hello guys, sorry for the late response.

To achieve this it would be necessary to be able to programatically manipulate the Reviewers input in the pull request creation/editing view.

From a little bit of experimentation and reverse engineering it didn't seem trivial to me at first glance, I would have to spend more time trying to figure it out. If anyone else would like to take a look and discuss it please feel free to do so!

**Comment by drexler** [comment]
@reyronald  i'll have some free time over the weekend to take a closer look at this too.

**Comment by reyronald** [comment]
@drexler awesome! Let me know how it goes!

**Comment by marco-tranzatto** [comment]
Great to hear that @drexler !
I also would like to know how it goes once if you managed to give it a try :)
